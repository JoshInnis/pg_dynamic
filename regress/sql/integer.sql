/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * Bigint Typecasting
 */
 SELECT ('1'::dynamic)::bigint;
 SELECT ('"1"'::dynamic)::bigint;
 SELECT ('1.0'::dynamic)::bigint;
 SELECT ('1::numeric'::dynamic)::bigint;

 SELECT (1::bigint)::dynamic;

SELECT dynamic_tobigint('1');
SELECT dynamic_tobigint('"1"');
SELECT dynamic_tobigint('1.0');
SELECT dynamic_tobigint('1::numeric');
SELECT dynamic_tobigint('true');
SELECT dynamic_tobigint('false');
SELECT dynamic_tobigint('null');
SELECT dynamic_tobigint('"2023-06-23 13:39:40.00"::timestamp');
SELECT dynamic_tobigint('"1997-12-17 07:37:16-06"::timestamptz');
SELECT dynamic_tobigint('"1997-12-17"::date');
SELECT dynamic_tobigint('"07:37:16-08"::time');
SELECT dynamic_tobigint('"07:37:16"::timetz');
SELECT dynamic_tobigint('"30 Seconds"::interval');
SELECT dynamic_tobigint('"192.168.1.5"::inet');
SELECT dynamic_tobigint('"192.168.1.5"::cidr');
SELECT dynamic_tobigint('[-1, 0, 1]');
SELECT dynamic_tobigint('{"bool":true, "null":null, "string":"string", "integer":1, "float":1.2, "arrayi":[-1,0,1], "arrayf":[-1.0, 0.0, 1.0], "object":{"bool":true, "null":null, "string":"string", "int":1, "float":8.0}}');

SELECT dynamic_tobigint('nan');
SELECT dynamic_tobigint('Infinity');
SELECT dynamic_tobigint('-Infinity');
SELECT dynamic_tobigint('inf');
SELECT dynamic_tobigint('-inf');

--
-- Operators
--
SELECT '10'::dynamic + '100'::dynamic;
SELECT '"60"'::dynamic + '100'::dynamic;
SELECT '1.0'::dynamic + '100'::dynamic;
SELECT '50::numeric'::dynamic + '100'::dynamic;
SELECT '-1'::dynamic + '100'::dynamic;

SELECT '10'::dynamic - '100'::dynamic;
SELECT '"60"'::dynamic - '100'::dynamic;
SELECT '1.0'::dynamic - '100'::dynamic;
SELECT '50::numeric'::dynamic - '100'::dynamic;
SELECT '-1'::dynamic - '100'::dynamic;


SELECT '10'::dynamic * '100'::dynamic;
SELECT '"60"'::dynamic * '100'::dynamic;
SELECT '1.0'::dynamic * '100'::dynamic;
SELECT '50::numeric'::dynamic * '100'::dynamic;
SELECT '-1'::dynamic * '100'::dynamic;

--
-- Functions
--
 SELECT abs('1'::dynamic);
 SELECT abs('"1"'::dynamic);
 SELECT abs('1.0'::dynamic);
 SELECT abs('1::numeric'::dynamic);
 SELECT abs('-1'::dynamic);
 SELECT abs('0'::dynamic);


 SELECT gcd('10'::dynamic, '100'::dynamic);
 SELECT gcd('"60"'::dynamic, '100'::dynamic);
 SELECT gcd('1.0'::dynamic, '100'::dynamic);
 SELECT gcd('50::numeric'::dynamic, '100'::dynamic);
 SELECT gcd('-1'::dynamic, '100'::dynamic);
 SELECT gcd('101'::dynamic, '100'::dynamic);

 SELECT lcm('10'::dynamic, '100'::dynamic);
 SELECT lcm('"60"'::dynamic, '100'::dynamic);
 SELECT lcm('1.0'::dynamic, '100'::dynamic);
 SELECT lcm('50::numeric'::dynamic, '100'::dynamic);
 SELECT lcm('-1'::dynamic, '100'::dynamic);
 SELECT lcm('101'::dynamic, '100'::dynamic);
