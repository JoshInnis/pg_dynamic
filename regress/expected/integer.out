/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
/*
 * Bigint Typecasting
 */
 SELECT ('1'::dynamic)::bigint;
 int8 
------
    1
(1 row)

 SELECT ('"1"'::dynamic)::bigint;
 int8 
------
    1
(1 row)

 SELECT ('1.0'::dynamic)::bigint;
 int8 
------
    1
(1 row)

 SELECT ('1::numeric'::dynamic)::bigint;
 int8 
------
    1
(1 row)

 SELECT (1::bigint)::dynamic;
 dynamic 
---------
 1
(1 row)

SELECT dynamic_tobigint('1');
 dynamic_tobigint 
------------------
 1
(1 row)

SELECT dynamic_tobigint('"1"');
 dynamic_tobigint 
------------------
 1
(1 row)

SELECT dynamic_tobigint('1.0');
 dynamic_tobigint 
------------------
 1
(1 row)

SELECT dynamic_tobigint('1::numeric');
 dynamic_tobigint 
------------------
 1
(1 row)

SELECT dynamic_tobigint('true');
ERROR:  cannot cast dynamic boolean to type int8
SELECT dynamic_tobigint('false');
ERROR:  cannot cast dynamic boolean to type int8
SELECT dynamic_tobigint('null');
ERROR:  cannot cast dynamic null to type int8
SELECT dynamic_tobigint('"2023-06-23 13:39:40.00"::timestamp');
ERROR:  cannot cast dynamic timestamp to type int8
SELECT dynamic_tobigint('"1997-12-17 07:37:16-06"::timestamptz');
ERROR:  cannot cast dynamic timestamptz to type int8
SELECT dynamic_tobigint('"1997-12-17"::date');
ERROR:  cannot cast dynamic date to type int8
SELECT dynamic_tobigint('"07:37:16-08"::time');
ERROR:  cannot cast dynamic time to type int8
SELECT dynamic_tobigint('"07:37:16"::timetz');
ERROR:  cannot cast dynamic timetz to type int8
SELECT dynamic_tobigint('"30 Seconds"::interval');
ERROR:  cannot cast dynamic interval to type int8
SELECT dynamic_tobigint('"192.168.1.5"::inet');
ERROR:  unknown dynamic type: 12 can't cast to int8
SELECT dynamic_tobigint('"192.168.1.5"::cidr');
ERROR:  unknown dynamic type: 13 can't cast to int8
SELECT dynamic_tobigint('[-1, 0, 1]');
ERROR:  cannot cast non-scalar dynamic to dynamic integer
SELECT dynamic_tobigint('{"bool":true, "null":null, "string":"string", "integer":1, "float":1.2, "arrayi":[-1,0,1], "arrayf":[-1.0, 0.0, 1.0], "object":{"bool":true, "null":null, "string":"string", "int":1, "float":8.0}}');
ERROR:  cannot cast non-scalar dynamic to dynamic integer
SELECT dynamic_tobigint('nan');
ERROR:  bigint out of range
SELECT dynamic_tobigint('Infinity');
ERROR:  bigint out of range
SELECT dynamic_tobigint('-Infinity');
ERROR:  bigint out of range
SELECT dynamic_tobigint('inf');
ERROR:  bigint out of range
SELECT dynamic_tobigint('-inf');
ERROR:  bigint out of range
--
-- Operators
--
SELECT '10'::dynamic + '100'::dynamic;
 ?column? 
----------
 110
(1 row)

SELECT '"60"'::dynamic + '100'::dynamic;
 ?column? 
----------
 160
(1 row)

SELECT '1.0'::dynamic + '100'::dynamic;
 ?column? 
----------
 101
(1 row)

SELECT '50::numeric'::dynamic + '100'::dynamic;
 ?column? 
----------
 150
(1 row)

SELECT '-1'::dynamic + '100'::dynamic;
 ?column? 
----------
 99
(1 row)

SELECT + '10'::dynamic;
 ?column? 
----------
 10
(1 row)

SELECT + '"60"'::dynamic;
 ?column? 
----------
 60
(1 row)

SELECT + '1.0'::dynamic;
 ?column? 
----------
 1
(1 row)

SELECT + '50::numeric'::dynamic;
 ?column? 
----------
 50
(1 row)

SELECT + '-1'::dynamic;
 ?column? 
----------
 -1
(1 row)

SELECT '10'::dynamic - '100'::dynamic;
 ?column? 
----------
 -90
(1 row)

SELECT '"60"'::dynamic - '100'::dynamic;
 ?column? 
----------
 -40
(1 row)

SELECT '1.0'::dynamic - '100'::dynamic;
 ?column? 
----------
 -99
(1 row)

SELECT '50::numeric'::dynamic - '100'::dynamic;
 ?column? 
----------
 -50
(1 row)

SELECT '-1'::dynamic - '100'::dynamic;
 ?column? 
----------
 -101
(1 row)

SELECT - '10'::dynamic;
 ?column? 
----------
 -10
(1 row)

SELECT - '"60"'::dynamic;
 ?column? 
----------
 -60
(1 row)

SELECT - '1.0'::dynamic;
 ?column? 
----------
 -1
(1 row)

SELECT - '50::numeric'::dynamic;
 ?column? 
----------
 -50
(1 row)

SELECT - '-1'::dynamic;
 ?column? 
----------
 1
(1 row)

SELECT '10'::dynamic * '100'::dynamic;
 ?column? 
----------
 1000
(1 row)

SELECT '"60"'::dynamic * '100'::dynamic;
 ?column? 
----------
 6000
(1 row)

SELECT '1.0'::dynamic * '100'::dynamic;
 ?column? 
----------
 100
(1 row)

SELECT '50::numeric'::dynamic * '100'::dynamic;
 ?column? 
----------
 5000
(1 row)

SELECT '-1'::dynamic * '100'::dynamic;
 ?column? 
----------
 -100
(1 row)

SELECT '1000'::dynamic / '100'::dynamic;
 ?column? 
----------
 10
(1 row)

SELECT '"600"'::dynamic / '100'::dynamic;
 ?column? 
----------
 6
(1 row)

SELECT '1000.0'::dynamic / '100'::dynamic;
 ?column? 
----------
 10
(1 row)

SELECT '5000::numeric'::dynamic / '100'::dynamic;
 ?column? 
----------
 50
(1 row)

SELECT '-10000'::dynamic / '100'::dynamic;
 ?column? 
----------
 -100
(1 row)

--
-- Functions
--
 SELECT abs('1'::dynamic);
 abs 
-----
 1
(1 row)

 SELECT abs('"1"'::dynamic);
 abs 
-----
 1
(1 row)

 SELECT abs('1.0'::dynamic);
 abs 
-----
 1
(1 row)

 SELECT abs('1::numeric'::dynamic);
 abs 
-----
 1
(1 row)

 SELECT abs('-1'::dynamic);
 abs 
-----
 1
(1 row)

 SELECT abs('0'::dynamic);
 abs 
-----
 0
(1 row)

 SELECT gcd('10'::dynamic, '100'::dynamic);
 gcd 
-----
 10
(1 row)

 SELECT gcd('"60"'::dynamic, '100'::dynamic);
 gcd 
-----
 20
(1 row)

 SELECT gcd('1.0'::dynamic, '100'::dynamic);
 gcd 
-----
 1
(1 row)

 SELECT gcd('50::numeric'::dynamic, '100'::dynamic);
 gcd 
-----
 50
(1 row)

 SELECT gcd('-1'::dynamic, '100'::dynamic);
 gcd 
-----
 1
(1 row)

 SELECT gcd('101'::dynamic, '100'::dynamic);
 gcd 
-----
 1
(1 row)

 SELECT lcm('10'::dynamic, '100'::dynamic);
 lcm 
-----
 100
(1 row)

 SELECT lcm('"60"'::dynamic, '100'::dynamic);
 lcm 
-----
 300
(1 row)

 SELECT lcm('1.0'::dynamic, '100'::dynamic);
 lcm 
-----
 100
(1 row)

 SELECT lcm('50::numeric'::dynamic, '100'::dynamic);
 lcm 
-----
 100
(1 row)

 SELECT lcm('-1'::dynamic, '100'::dynamic);
 lcm 
-----
 100
(1 row)

 SELECT lcm('101'::dynamic, '100'::dynamic);
  lcm  
-------
 10100
(1 row)

